source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/AdaptiveCardRenderer.java

```diff
@@ -60,17 +62,42 @@ public RenderedAdaptiveCard render(
         return render(context, fragmentManager, adaptiveCard, cardActionHandler, null, hostConfig);
     }
 
+    public RenderedAdaptiveCard render(Context context,
+                                       FragmentManager fragmentManager,
+                                       AdaptiveCard adaptiveCard,
+                                       ICardActionHandler cardActionHandler,
+                                       @Nullable IOverflowActionRenderer overflowActionRenderer,
+                                       @Nullable IChoicesResolver choicesResolver,
+                                       @Nullable ActivityResultRegistry activityResultRegistry)
+    {
+        return render(context, fragmentManager, adaptiveCard, cardActionHandler, overflowActionRenderer, choicesResolver, activityResultRegistry, defaultHostConfig);
+    }
+
+    public RenderedAdaptiveCard render(
+        Context context,
+        FragmentManager fragmentManager,
+        AdaptiveCard adaptiveCard,
+        ICardActionHandler cardActionHandler,
+        @Nullable IOverflowActionRenderer overflowActionRenderer,
+        HostConfig hostConfig)
+    {
+        return render(context, fragmentManager, adaptiveCard, cardActionHandler, overflowActionRenderer, null, null, hostConfig);
+    }
+
     public RenderedAdaptiveCard render(
         Context context,
         FragmentManager fragmentManager,
         AdaptiveCard adaptiveCard,
         ICardActionHandler cardActionHandler,
         @Nullable IOverflowActionRenderer overflowActionRenderer,
+        @Nullable IChoicesResolver choicesResolver,
+        @Nullable ActivityResultRegistry activityResultRegistry,
         HostConfig hostConfig)
     {
         RenderedAdaptiveCard result = new RenderedAdaptiveCard(adaptiveCard);
         CardRendererRegistration.getInstance().registerOverflowActionRenderer(overflowActionRenderer);
-        View cardView = internalRender(result, context, fragmentManager, adaptiveCard, cardActionHandler, hostConfig, false, View.NO_ID);
+        CardRendererRegistration.getInstance().registerActivityResultRegistry(activityResultRegistry);
+        View cardView = internalRender(result, context, fragmentManager, adaptiveCard, cardActionHandler, choicesResolver, hostConfig, false, View.NO_ID);
```

Comment:
>choicesResolver [](http://example.com/codeflow?start=106&length=15)

why is choicesResolver not registered as the other two new parameters in this method?

Reply:
@almedina-ms - Let's think of a scenario when there are multiple cards in 1 activity/screen (teams or outlook).

In such a case if we register a paramter then it shall be common for all cards in the activity/screen. Other parameters such as activity register or custom overflow menu renderer are common for all such cards. 

But when we talk about action handling or fetching dynamic choices, it can be different for different cards. For example there are two cards in an activity card 1 and card 2. Now actions clicked on card 1 can be handled differently than how there are handled on card 2 (in case of teams invokes are sent to different bots who have sent the card).

That's why ICardActionHanlder and IChoicesResolver are per card and need to be sent as parameters in the render() method. Since this is an interface method, we cannot overloading is not possible which makes this a breaking change. But as I explained earlier, this is a required change.

Let me know if you have any further questions

Reply:
I have introduced a `ChannelAdaptor` which is responsible for any host/SDK communication. Currently it will contain `ICardActionHandler` and `IChoicesResolver`.

This will enable us to prevent any breaking changes in the future if new host/SDK interactions are introduced.